Add endpoint to retrieve all the current prices

yabirgb opened the pull request:
- Removes the endpoint to query NFTs prices since as discussed with @lukicenturi this can be retrieved from the NFTs balances where the same query is executed
- Adds for the get verb the possibility to query all the manual current prices that the user has added

LefterisJP commented on rotkehlchen/api/rest.py:
```diff
@@ -3787,13 +3787,17 @@ def get_nfts_balances(self, async_query: bool, ignore_cache: bool) -> Response:
-            method='get_nfts_with_price',
-            query_specific_balances_before=None,
-        )
+    def get_manual_current_prices(self) -> Response:
```
Shouldn't this be a paginated one? Perhaps an overkill for now but in the future I can see this becoming too long. Can also go in another PR to unblock Luki faster.

yabirgb replied:
I asked luki about pagination in this endpoint and told me to keep it unpaginated. Should I change it? I also wonder the same

lukicenturi replied:
if you want to make it paginated, I need backend to support filter by asset too 😛

LefterisJP replied:
I mean this is easy I guess. The point is to be future-proof. How should this endpoint be for the frontend? @kelsos ?

kelsos replied:
I am not sure how pagination would help in this cache, for future-proofing we would need to change the approach we work with prices in the frontend.

Currently, the prices for owned assets are also cached in the state.

LefterisJP replied:
okay but the idea is to go towards pagination everything and not cache stuff in the frontend, right?

LefterisJP commented on rotkehlchen/tests/api/test_current_assets_price.py:
```diff
@@ -305,3 +305,47 @@ def test_inquirer_oracles_affect_manual_price(inquirer):
+        ),
+    )
+    result = assert_proper_response_with_result(response)
+    expected_response = [
```
shouldn't you also check that the first ETH/EUR price becomes a historical price?

yabirgb replied:
That should be tested in a different test when adding the manual current prices. I see that is not the case so would add a test for that

LefterisJP commented on rotkehlchen/globaldb/handler.py:
```diff
@@ -1055,6 +1055,28 @@ def get_manual_current_price(asset: Asset) -> Optional[Tuple[Asset, Price]]:
+        with GlobalDBHandler().conn.read_ctx() as read_cursor:
+            result = read_cursor.execute(query, bindings)
+            return [
+                (Asset(entry[0]), Asset(entry[1]), deserialize_price(entry[2]))
```
Do you really need to deserialize to `Asset` here? Isn't this result just going out of the API, which means it just gets turned to an identifer again?

yabirgb replied:
I made it follow the pattern in the handler of deserializing when is needed but yes... is getting the asset to only return the identifier later. If is okey with you will remove that

yabirgb commented:
@LefterisJP I've changed the deserialization of the Asset as you mentioned.

Also @lukicenturi and I checked the code and seems that the nft balacnes endpoint even when queried with `ignore_cache=False` it queries the uniswap v3 module so I've added again the nft endpoint that we had under a different resource. This is needed because provided only the prices for nfts queried.

We tough it was possible to get it from the balances resource but seems to be slower

LefterisJP commented on docs/api.rst:
```diff
@@ -1140,6 +1141,57 @@ Get current price and custom price for assets
+
+.. http:post:: /api/(version)/assets/prices/current/all
+
+   Retrieve all the manual current prices stored in the database.
```
```suggestion
   Retrieve all the manual latest prices stored in the database.
```

LefterisJP commented on docs/api.rst:
```diff
@@ -1140,6 +1141,57 @@ Get current price and custom price for assets
+
+   .. http:example:: curl wget httpie python-requests
+
+      POST /api/1/assets/prices/current HTTP/1.1
```
```suggestion
      POST /api/1/assets/prices/latest HTTP/1.1
```

Generally change current to latest everywhere

LefterisJP commented on docs/api.rst:
```diff
@@ -1140,6 +1141,57 @@ Get current price and custom price for assets
    :statuscode 502: An external service used in the query such as cryptocompare/coingecko could not be reached or returned unexpected response.
 
+
+Get current price and custom price for assets
```
```suggestion
Get current price and custom price for assets
```

Is this current and custom? Or just always latest price? Or only latest manual? The title is confusin

prettyirrelevant commented on rotkehlchen/inquirer.py:
```diff
@@ -412,55 +430,117 @@ def _query_oracle_instances(
         """
         if asset == A_USD:
-            return Price(ONE)
+            return Price(ONE), CurrentPriceOracle.FIAT
```
Should this be FIAT? Maybe None?

prettyirrelevant replied:
This shouldn't be `CurrentPriceOracle.FIAT` right? Given that it was not gotten from any oracle.

prettyirrelevant commented on rotkehlchen/inquirer.py:
```diff
@@ -412,55 +430,117 @@ def _query_oracle_instances(
+        return price
+
+    @staticmethod
+    def find_usd_price_and_oracles(
```
```suggestion
    def find_usd_price_and_oracle(
```

LefterisJP replied:
+1

LefterisJP replied:
This is still not addressed ...

prettyirrelevant reviewed (commented):
@yabirgb left you some comments

LefterisJP commented on rotkehlchen/chain/ethereum/modules/nfts.py:
```diff
@@ -221,44 +221,52 @@ def get_balances(
-            except UnknownAsset:
-                log.error(f'Unknown asset {to_asset_id} in custom nft price DB table. Ignoring.')
-                continue
+    def get_nfts_with_price(self, identifier: Optional[str] = None) -> List[Dict[str, Any]]:
```
add a docstring

yabirgb replied:
This is not new code. Not sure why github shows it in the diff like this :/

LefterisJP replied:
because code moved. Just add the docstring

LefterisJP commented on rotkehlchen/chain/ethereum/modules/nfts.py:
```diff
@@ -221,44 +221,52 @@ def get_balances(
+                            f'price api call. Ignoring.',
+                        )
+                        continue
+                    usd_price = to_asset_usd_price * FVal(entry[1])
```
shouldn't you do a sanity check that entry[1] can be deserialized to an fval for here? Or we always check at insertion so we assume is always good? Probably the latter eh?

yabirgb replied:
yes, the latter one

LefterisJP commented on rotkehlchen/globaldb/handler.py:
```diff
@@ -1197,7 +1197,29 @@ def get_manual_current_price(asset: Asset) -> Optional[Tuple[Asset, Price]]:
+    def get_all_manual_latest_prices(
+            from_asset: Optional[Asset] = None,
+            to_asset: Optional[Asset] = None,
+    ) -> List[Tuple[Asset, Asset, Price]]:
```
types are wrong here. It should be a tuple of 3 strings

LefterisJP commented on rotkehlchen/globaldb/handler.py:
```diff
@@ -1197,7 +1197,29 @@ def get_manual_current_price(asset: Asset) -> Optional[Tuple[Asset, Price]]:
+        with GlobalDBHandler().conn.read_ctx() as read_cursor:
+            result = read_cursor.execute(query, bindings)
+            return [
+                (entry[0], entry[1], deserialize_price(entry[2]))
```
you do not need to do this at all. Just do a fetchall(). This is already 3 string values in each entry and this function is only used at one place, where you serialize to string to return through the api. So this is not needed at all

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -167,52 +169,65 @@ def get_underlying_asset_price(token: EvmToken) -> Optional[Price]:
+            usd_price += underlying_asset_price * underlying_token.weight
+            oracles.append(oracle)
+
+        # Set as oracle the most common oracle of the underlying assets
```
this is an overkill. Why not just keep the oracle you had from outside the if.

yabirgb replied:
Because in this case different oracles were queried

LefterisJP replied:
I said this is an overkill.

This information is not needed. Can you think of a situation where this is actually needed? Please describe it to me and explain why I would want such detail.

Just set price oracle blockchain and be done with it.
Then all this extra logic of appending the oracle and then doing a `max()` can go away.

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -365,7 +381,7 @@ def _query_oracle_instances(
             to_asset: Asset,
             skip_onchain: bool = False,
-    ) -> Price:
+    ) -> Tuple[Price, Optional[CurrentPriceOracle]]:
```
The Oracle should not be optional in the return here

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -412,55 +430,117 @@ def _query_oracle_instances(
         """
         if from_asset == to_asset:
-            return Price(FVal('1'))
+            return Price(FVal('1')), CurrentPriceOracle.MANUALCURRENT
```
Use the `ONE` constant

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -412,55 +430,117 @@ def _query_oracle_instances(
+            to_asset: Asset,
+            ignore_cache: bool = False,
+            skip_onchain: bool = False,
+    ) -> Tuple[Price, Optional[CurrentPriceOracle]]:
```
should not be optional

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -412,55 +430,117 @@ def _query_oracle_instances(
+            asset: Asset,
+            ignore_cache: bool = False,
+            skip_onchain: bool = False,
+    ) -> Tuple[Price, Optional[CurrentPriceOracle]]:
```
should not be optional

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -598,7 +680,7 @@ def find_curve_pool_price(
         prices = []
         for token in tokens:
-            price = self.find_usd_price(token)
+            price, _ = self.find_usd_price_and_oracles(token)
```
why not use the wrapper you made here?

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -690,7 +772,7 @@ def find_yearn_price(
 
         underlying_token = EvmToken(ethaddress_to_identifier(maybe_underlying_token[0].address))
-        underlying_token_price = self.find_usd_price(underlying_token)
+        underlying_token_price, _ = self.find_usd_price_and_oracles(underlying_token)
```
can't you use the wrapper you made?

yabirgb replied:
Yes, this is the terrible idea I had before creating the wrapper

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -770,7 +853,7 @@ def _query_fiat_pair(base: Asset, quote: Asset) -> Price:
         """
         if base == quote:
-            return Price(FVal('1'))
+            return Price(FVal('1')), CurrentPriceOracle.FIAT
```
use the `ONE` constant

yabirgb commented on rotkehlchen/api/rest.py:
```diff
@@ -3819,6 +3835,25 @@ def get_nfts_balances(self, async_query: bool, ignore_cache: bool) -> Response:
+                'to_asset': price_entry[1],
+                'price': price_entry[2],
+            })
+        processed_result = process_result_list(prices_information)
```
Why? it contains asset that need to be serialized

LefterisJP commented on docs/api.rst:
```diff
@@ -1033,7 +1033,7 @@ Query the result of an ongoing backend task
-.. http:post:: /api/(version)/assets/prices/current
+.. http:post:: /api/(version)/assets/prices/latest
 
    Querying this endpoint with a list of assets and a target asset will return an object with the the price of the assets in the target asset currency. Providing an empty list or no target asset is an error.
```
You did not do this. Why resolve it? The description of the endpoint is still unchanged.

It says:

> Querying this endpoint with a list of assets and a target asset will return an object with the the price of the assets in the target asset currency. Providing an empty list or no target asset is an error.

Which does not mention what an oracle is or that it's returned.

LefterisJP commented on rotkehlchen/tests/api/test_current_assets_price.py:
```diff
@@ -303,3 +304,98 @@ def test_inquirer_oracles_affect_manual_price(inquirer):
+
+@pytest.mark.parametrize('should_mock_current_price_queries', [False])
+def test_get_all_current_prices(rotkehlchen_api_server):
+    """Test that the endpoint to fetch all manual imput returns the correct prices results"""
```
```suggestion
    """Test that the endpoint to fetch all manual input returns the correct prices results"""
```

LefterisJP reviewed (commented):
Some minor comments. But man please never resolve something without addressing it and don't skip review comments. It adds so much more time to the entire process

LefterisJP commented on docs/api.rst:
```diff
@@ -1030,12 +1030,12 @@ Query the result of an ongoing backend task
+.. http:post:: /api/(version)/assets/prices/latest
 
-   Querying this endpoint with a list of assets and a target asset will return an object with the the price of the assets in the target asset currency. Providing an empty list or no target asset is an error.
+   Querying this endpoint with a list of assets and a target asset will return a mapping of assets to their prices in the target asset and an integer representing the oracle the price was gotten from. Providing an empty list or no target asset is an error.
```
```suggestion
   Querying this endpoint with a list of assets and a target asset will return a mapping of assets to their prices in the target asset and an integer representing the oracle the price was retrieved from. Providing an empty list or no target asset is an error.
```

LefterisJP commented on rotkehlchen/inquirer.py:
```diff
@@ -467,15 +481,50 @@ def find_price(
+        return price
+
+    @staticmethod
+    def find_price_and_oracles(
```
this plural here should also have been addressed... but at this point I am just tired of repeating myself in reviews and don't care. Dissapointed ...

LefterisJP reviewed (approved):
Have 2 more comments but at this point I am just tired and will merge as is. Please remember them in other PRs and address them.
